### PR TITLE
Overhaul executor <-> redis interface

### PR DESCRIPTION
**Pull request by SimonBoothroyd**
## Description

This PR:

* abstracts the coordinator <-> redis interface into own its module
* fixes a gateway hanging error when processing many tasks due to async congestion
* adds safety checks to ensure version compatibility
* adds user-friendly messages when possible redis issues are encountered

## Status
- [X] Ready to go

### CODE REVIEW

File: openff/bespokefit/executor/services/_settings.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ class Settings(BaseSettings):
     BEFLOW_GATEWAY_LOG_LEVEL: str = "error"
 
     BEFLOW_REDIS_ADDRESS: str = "localhost"
-    BEFLOW_REDIS_PORT: int = 6379
+    BEFLOW_REDIS_PORT: int = 6363
```
Comment (SimonBoothroyd) [inline openff/bespokefit/executor/services/_settings.py:31]: By changing the default port hopefully we're less likely to accidentally connect to someones running redis server unintentionally!

Comment (jthorton) [reply openff/bespokefit/executor/services/_settings.py:31]: Great idea, this should be much safer.

File: openff/bespokefit/executor/services/_settings.py [hunk 2]
```diff
@@ -38,6 +38,8 @@ class Settings(BaseSettings):
     BEFLOW_COORDINATOR_WORKER = (
         "openff.bespokefit.executor.services.coordinator.worker:"
     )
+    BEFLOW_COORDINATOR_MAX_UPDATE_INTERVAL: float = 5.0
```
[the remark this answers is not in the document]
Comment (SimonBoothroyd) [reply openff/bespokefit/executor/services/_settings.py:41]: Great question! We should be safe here now that we have [this line added](https://github.com/openforcefield/openff-bespokefit/blob/990238fc089d081c6fd1ddcea5a9642b45d09042/openff/bespokefit/executor/services/coordinator/worker.py#L87), this mainly just lets users tweak how much CPU the coordinator will use on average when only a few molecules are being fit!

File: openff/bespokefit/executor/services/coordinator/worker.py [hunk 3]
```diff
@@ -1,78 +1,117 @@
...
+
+            n_running_tasks = get_n_tasks(TaskStatus.running)
+            n_tasks_to_queue = min(
+                settings.BEFLOW_COORDINATOR_MAX_RUNNING_TASKS - n_running_tasks,
+                get_n_tasks(TaskStatus.waiting),
+            )
+
+            for _ in range(n_tasks_to_queue):
+
+                push_task_status(
```
Comment (jthorton) [inline openff/bespokefit/executor/services/coordinator/worker.py:106]: This is fantastic, great idea to use separate queues!

File: openff/bespokefit/executor/utilities/redis.py [hunk 4]
```diff
@@ -5,12 +5,87 @@
...
+
+
+def connect_to_redis(
+    host: str, port: int, db: int, validate: bool = True
+) -> redis.Redis:
+    """Connects to a redis server using the specified settings."""
+
+    connection_key = (host, port, db, validate)
+
+    if connection_key in __CONNECTION_POOL:
```
Comment (jthorton) [inline openff/bespokefit/executor/utilities/redis.py:56]: Is there any danger with multiple workers using the same connection object to redis?

File: openff/bespokefit/executor/utilities/redis.py [hunk 5]
```diff
@@ -93,4 +172,16 @@ def launch_redis(
...
+    try:
+        connect_to_redis("localhost", port, 0, validate=True)
+    except RedisNotConfiguredError:
+
+        if redis_save_exists:
+            raise
+
+        connection = connect_to_redis("localhost", port, 0, validate=False)
+        connection.set(
+            "openff-bespokefit:redis-version", expected_redis_config_version()
```
Comment (jthorton) [inline openff/bespokefit/executor/utilities/redis.py:184]: Great idea to stamp a version on the database as well!

**Comment by jthorton** [comment]
This looks great, the new coordinator storage file looks like the start of a client object that we could add later if there is interest! I also think it might be worth extending the `executor list` command to take a status so users can check on specific queues or all tasks if they pass `all` as listing every task for large databases may take some time?

File: openff/bespokefit/executor/utilities/redis.py (hunk 4 above)
Comment (SimonBoothroyd) [reply openff/bespokefit/executor/utilities/redis.py:56]: I'm not 100% certain, but I'm pretty sure it won't affect anything as the connection shouldn't really be doing anything other than pinging the redis server!

**Comment by SimonBoothroyd** [comment]
> I also think it might be worth extending the executor list command to take a status so users can check on specific queues or all tasks if they pass all as listing every task for large databases may take some time?

Definitely, I can open a separate PR for this once this is merged so this one doesn't get even bigger!

**Pull request merged**